src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,329 @@
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.IdentityModel.Logging;
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    /// <summary>
+    /// This is an LRU cache implementation that relies on an event queue rather than locking to achieve thread safety.
+    /// This approach has been decided on in order to optimize the performance of the get and set operations on the cache.
+    /// This cache contains a doubly linked list in order to maintain LRU order, as well as a dictionary (map) to keep track of
+    /// keys and expiration times. The linked list (a structure which is not thread-safe) is NEVER modified directly inside
+    /// an API call (e.g. get, set, remove); it is only ever modified sequentially by a background thread. On the other hand,
+    /// the map is a <see cref="ConcurrentDictionary{TKey, TValue}"/> which may be modified directly inside an API call or
+    /// through eventual processing of the event queue. This implementation relies on the principle of 'eventual consistency':
+    /// though the map and it's corresponding linked list may be out of sync at any given point in time, they will eventually line up.
+    /// </summary>
+    /// <typeparam name="TKey">The key type to be used by the cache.</typeparam>
+    /// <typeparam name="TValue">The value type to be used by the cache</typeparam>
+    internal class EventBasedLRUCache<TKey, TValue> : IDisposable
+    {
+        private readonly int _capacity;
+        // The percentage of the cache to be removed when _maxCapacityPercentage is reached.
+        private readonly double _compactionPercentage = .20;
+        private LinkedList<LRUCacheItem<TKey, TValue>> _doubleLinkedList = new LinkedList<LRUCacheItem<TKey, TValue>>();
+        private BlockingCollection<Action> _eventQueue = new BlockingCollection<Action>();
+        private Task _eventQueueTask;
+        private ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>> _map;
+        // When the current cache size gets to this percentage of _capacity, _compactionPercentage% of the cache will be removed.
+        private readonly double _maxCapacityPercentage = .95;
+        private bool _disposed = false;
+
+        internal EventBasedLRUCache(int capacity, TaskCreationOptions options = TaskCreationOptions.LongRunning, IEqualityComparer<TKey> comparer = null)
+        {
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _eventQueueTask = new Task(OnStart, options);
+            _eventQueueTask.Start();
+            _ = RemoveExpiredValuesPeriodically(TimeSpan.FromMinutes(5));
```

Comment:
For the scenario of SignatureProviders, it may not be necessary to schedule removal of expired items. 
We could introduce a variable to control removal and the interval in CryptoProviderCacheOptions.